src/res/mod.rs

```diff
@@ -244,124 +257,293 @@ impl Resources {
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
     /// Fetches the resource with the specified type `T` mutably.
     ///
     /// Please see `fetch` for details.
-    pub fn fetch_mut<T>(&self) -> FetchMut<T>
+    pub fn fetch_mut<T, F, P>(&self, with: F) -> FetchMut<T, P>
     where
         T: Resource,
+        F: FnOnce() -> T,
     {
-        self.try_fetch_mut().expect(RESOURCE_NOT_FOUND)
+        let res_id = ResourceId::new::<T>();
+        let r = self.resources
+            .get(&res_id)
+            .unwrap_or_else(|| self.def_res::<T>(res_id, with()));
+
+        FetchMut {
+            inner: r.borrow_mut(),
+            phantom: PhantomData,
+        }
     }
 
-    /// Like `fetch_mut`, but returns an `Option` instead of panicking in the case of the resource
-    /// being accessed mutably.
-    pub fn try_fetch_mut<T>(&self) -> Option<FetchMut<T>>
+    /// Like `fetch_mut`, but returns an `Option` instead of inserting a default value
+    /// in case the resource does not exist.
+    pub fn try_fetch_mut<T, P>(&self) -> Option<FetchMut<T, P>>
     where
         T: Resource,
     {
-        self.try_fetch_internal(TypeId::of::<T>()).map(|r| {
-            FetchMut {
+        let res_id = ResourceId::new::<T>();
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| FetchMut {
                 inner: r.borrow_mut(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
-    fn try_fetch_internal(&self, id: TypeId) -> Option<&TrustCell<Box<Resource>>> {
-        self.resources.get(&ResourceId(id))
+    fn def_res<T>(&self, res_id: ResourceId, value: T) -> &TrustCell<Box<Resource>>
+    where
+        T: Resource,
+    {
+        let mut new = self.new.lock();
+        let b: &mut Box<_> = new.entry(res_id)
+            .or_insert_with(|| Box::new(TrustCell::new(Box::new(value) as Box<Resource>)));
+
+        Resources::box_to_cell_ref(b)
+    }
+
+    fn get_res<T>(&self, res_id: &ResourceId) -> Option<&TrustCell<Box<Resource>>>
+    where
+        T: Resource,
+    {
+        let new = self.new.lock();
+        let b: &Box<_> = match new.get(&res_id) {
+            Some(x) => x,
+            None => return None,
+        };
+
+        Some(Resources::box_to_cell_ref(b))
+    }
+
+    fn box_to_cell_ref<'a, 'b>(
+        b: &'a Box<TrustCell<Box<Resource>>>,
+    ) -> &'b TrustCell<Box<Resource>> {
+        unsafe {
+            // This is correct because the returned value can only live until the mutable borrow
+            // of `Resources`. This `Box` also lives at least until another mutable borrow,
+            // because the only way to drop it is calling `maintain` (which borrows `Resources`
+            // mutably).
+            // The raw pointer of a `Box` stays stable and valid as long as the `Box` doesn't
+            // get dropped.
+
```

Comment:
I don't think the safety of this unsafe block has anything to do with mutable borrow of `Resources` as these boxes are inside `Mutex`. It has more to do with the fact that only place we remove stuff from `new` is the `maintain` method.

Reply:
Ah now I realise what you mean by that. It's kinda subtle.

Reply:
Do you have a suggestion for how to improve this explanation? I usually don't write a big block of explanation for `unsafe` usage, but I thought this looks rather suspicious at first so I wanted to provide a comment.

Reply:
> Also the box_to_cell_ref is unsafe because it doesn't constrain the return lifetime to anything.

Yeah, that's the point. It'll just keep `res` borrowed, and as long as `res` is borrowed the returned reference is also valid.

Reply:
so this is impossible?:
```rust
let x = box_to_cell_ref(Box::new(TrustCell::new(Box::new(value) as Box<Resource>));
x.borrow()
```

Reply:
This is only safe combined to the lifetime bound it gets assigned by `def_res` or `get_res`. It'll bind the `TrustCell` to `'self`. The non-locality of unsafety is a big illusion :P

Reply:
"Because `Box` is allocated in heap it's pointer stays stable and only way that reference created here is invalidated is by dropping it.
Because `maintain` method is the only place that boxes in `new` are dropped reference created here must not survive after next maintain call.
This is ensured by the fact that `maintain` method requires mutable reference which enforces that all shared references are gone."

Reply:
Well you can make it take `self` as parameter and bind it to the return lifetime.

Reply:
Taking your text with small grammar fixes, thanks!